lookup/src/logup.rs

```diff
@@ -352,4 +376,107 @@ impl LookupGadget for LogUpGadget {
 
         deg_denom_constr.max(deg_num)
     }
+
+    fn generate_permutation<SC: StarkGenericConfig>(
+        &self,
+        main: &RowMajorMatrix<Val<SC>>,
+        public_values: &[Val<SC>],
+        lookups: &[Lookup<Val<SC>>],
+        lookup_data: &mut [LookupData<SC::Challenge>],
+        permutation_challenges: &[SC::Challenge],
+    ) -> RowMajorMatrix<SC::Challenge> {
+        let height = main.height();
+        let width = self.num_aux_cols() * lookups.len();
+
+        // Validate challenge count matches number of lookups.
+        debug_assert_eq!(
+            permutation_challenges.len(),
+            lookups.len() * self.num_challenges(),
+            "perm challenge count must be per-lookup"
+        );
+
+        // Enforce uniqueness of auxiliary column indices across lookups.
+        #[cfg(debug_assertions)]
+        {
+            let mut seen = BTreeSet::new();
+            for ctx in lookups {
+                let a = ctx.columns[0];
+                if !seen.insert(a) {
+                    panic!("duplicate aux column index {a} across lookups");
+                }
+            }
+        }
+
+        let mut aux_trace = vec![SC::Challenge::ZERO; height * width];
+        for i in 0..height {
+            let local_main_row = main.row_slice(i).unwrap();
+            let next_main_row = main.row_slice((i + 1) % height).unwrap();
+
+            let main_rows = VerticalPair::new(
+                RowMajorMatrixView::new_row(&local_main_row),
+                RowMajorMatrixView::new_row(&next_main_row),
+            );
+
+            let row_builder: LookupTraceBuilder<'_, SC> =
+                LookupTraceBuilder::new(main_rows, public_values, permutation_challenges.to_vec());
```

Comment:
I may be misunderstanding something here, but I don't think the `row` field is ever updated, hence keeping its initial value of 0, which conflicts with how methods like `is_last_row` / `is_first_row` are defined internally, when using this `row_builder` in eval functions?